### PR TITLE
Make AddHostedService register singletons

### PR DESCRIPTION
**Pull request by Tratcher**
https://github.com/aspnet/Extensions/issues/553

For the host it doesn't matter if IHostedServices are transient or singletons, they're only resolved once. However we've had several reports of users that also want to inject their hosted service into other components in order to interact with it.

Also adding an sln.

### CODE REVIEW

**Comment by dotnetchris** [comment]
Yes, hosted services should be viewed as 'servers' and a very common need of a server is to be able to expose status information about its health and state. With these currently registered as transient, it's impossible to access this information without using global variables to marshal state.

File: src/Hosting/Abstractions/src/ServiceCollectionHostedServiceExtensions.cs [hunk 1]
```diff
@@ -15,6 +15,6 @@ public static class ServiceCollectionHostedServiceExtensions
         /// <returns>The original <see cref="IServiceCollection"/>.</returns>
         public static IServiceCollection AddHostedService<THostedService>(this IServiceCollection services)
             where THostedService : class, IHostedService
-            => services.AddTransient<IHostedService, THostedService>();
+            => services.AddSingleton<IHostedService, THostedService>();
```
Comment (pakrym) [inline src/Hosting/Abstractions/src/ServiceCollectionHostedServiceExtensions.cs]: Add a test.

**Comment by davidfowl** [comment]
@dotnerchris you mean you want to use the hosted service as a global variable instead of another object....

**Comment by Tratcher** [comment]
@sguryev right. If it were Transient then the host and app would get difference instances and StartCalled would be false. Scoped wouldn't resolve at all, there is no scope established at this point.

**Comment by dotnetchris** [comment]
> @dotnetchris you mean you want to use the hosted service as a global variable instead of another object....

If you're making the pedantic argument that singleton and global are synonymous, sure.

**Comment by davidfowl** [comment]
> If you're making the pedantic argument that singleton and global are synonymous, sure.

No, I'm saying that you just want to use 1 global variable instead of 2. That's what this change enables.

**Comment by dotnetchris** [comment]
> > If you're making the pedantic argument that singleton and global are synonymous, sure.
> 
> No, I'm saying that you just want to use 1 global variable instead of 2. That's what this change enables.

In well designed adherence to encapsulation / open-closed principle. As opposed to side effecting global state in violation of _every_ design principle.

**Comment by Tratcher** [comment]
@billpoole-mi no change for IDisposable services. Note the host is single use, you can't call Start and Stop multiple times. All disposable services are disposed when you dispose the host, not when you call Stop.

File: src/Hosting/Hosting/test/HostTests.cs [hunk 2]
```diff
@@ -159,6 +159,27 @@ public async Task HostedServiceStartNotCalledIfHostNotStarted()
...
+        public async Task HostedServiceRegisteredAsSingletons()
+        {
+            using (var host = CreateBuilder()
+                   .ConfigureServices((hostContext, services) =>
+                   {
+                       services.AddHostedService<TestHostedService>();
+                   })
+                   .Build())
+            {
+                var svc = (TestHostedService)host.Services.GetRequiredService<IHostedService>();
```
Comment (AgentFire) [inline src/Hosting/Hosting/test/HostTests.cs:172]: Call `GetService` at elast twice to prove the instance is indeed the same?

Comment (dotnetchris) [reply src/Hosting/Hosting/test/HostTests.cs:172]: +1 and ensure reference equality of the 2nd result

Comment (davidfowl) [reply src/Hosting/Hosting/test/HostTests.cs:172]: Send a pull request?

Comment (sguryev) [reply src/Hosting/Hosting/test/HostTests.cs:172]: I have asked about it here: https://github.com/dotnet/extensions/pull/552#commitcomment-31388021
Answers are below.

**Pull request merged**